src/tools/save_slice_model.ts

```diff
@@ -0,0 +1,318 @@
+
+import { formatDecodeError, formatErrorForMcpTool } from "../lib/error"
+import { tool } from "../lib/mcp"
+import { trackSentryError } from "../lib/sentry"
+import {
+	initializeSliceMachineManager,
+	parseSliceMachineConfig,
+} from "../lib/sliceMachine"
+import { SharedSlice } from "@prismicio/types-internal/lib/customtypes"
+
+import { telemetryClient } from "../server"
+
+export const save_slice_model = tool(
+	"save_slice_model",
+	`PURPOSE: Creates or updates a Prismic slice in your project with a given valid model, performing local changes to the slice library.
+
+USAGE: Use to validate and create/update a slice with a given JSON model.
+
+RETURNS: Success confirmation or detailed validation errors if the model is invalid.`,
+	z.object({
+		sliceMachineConfigAbsolutePath: z
+			.string()
+			.describe(
+				"Absolute path to the slice library directory where slices live (from slicemachine.config.json)",
+			),
+		isNewSlice: z
+			.boolean()
+			.describe(
+				"Whether this is a new slice creation (true) or updating existing slice (false)",
+			),
+		sliceAbsolutePath: z
+			.string()
+			.describe(
+				`Absolute path to the directory of the slice to be created/updated`,
+			),
+		sliceModel: z
+			.record(z.string(), z.unknown())
+			.describe("The JSON model structure of the slice to be created/updated"),
+	}).shape,
+	async (args) => {
+		try {
+			const {
+				sliceMachineConfigAbsolutePath,
+				sliceAbsolutePath,
+				isNewSlice,
+				sliceModel: modelRaw,
+			} = args
+
+			const modelExists = existsSync(joinPath(sliceAbsolutePath, "model.json"))
+			if (isNewSlice && modelExists) {
+				throw new Error(
+					`Trying to create a new slice that already exists at ${sliceAbsolutePath}.`,
+				)
+			}
+			if (!isNewSlice && !modelExists) {
+				throw new Error(
+					`Trying to update a slice model that does not exist at ${sliceAbsolutePath}.`,
+				)
+			}
+
+			const sliceName = basename(sliceAbsolutePath)
+
+			try {
+				telemetryClient.track({
+					event: "MCP Tool - Save slice model",
+					sliceMachineConfigAbsolutePath,
+					properties: { sliceName, isNewSlice, sliceAbsolutePath },
+				})
+			} catch (error) {
+				// noop, we don't wanna block the tool call if the tracking fails
+				if (process.env.PRISMIC_DEBUG) {
+					console.error(
+						"Error while tracking 'save_slice_model' tool call:",
+						error,
+					)
+				}
+			}
+
+			let libraryID: string | undefined
+
+			try {
+				const projectRoot = dirname(sliceMachineConfigAbsolutePath)
+				const inputSliceLibraryPath = dirname(sliceAbsolutePath)
+				const smConfig = parseSliceMachineConfig(sliceMachineConfigAbsolutePath)
+
+				const resolvedAbsoluteLibraryPaths: string[] = []
+
+				for (const libraryPath of smConfig.libraries) {
+					const absPath = resolvePath(projectRoot, libraryPath)
+
+					resolvedAbsoluteLibraryPaths.push(absPath)
+
+					if (absPath === inputSliceLibraryPath) {
+						libraryID = libraryPath
+					}
+				}
```

Comment:
Did not reuse (like `getResolvedLibraries`) any util here because I need both the matching `libraryID` and the resolved libraries.

Reply:
i would have called the common function, then just use `find` to find the one you're looking for